File: LeetCode/Medium/1865-Finding_Pairs_With_A_Certain_Sum/1865-FindingPairsWithACertainSum.cpp
class FindSumPairs
{
private:
    vector<int> nums2;
    vector<int> keys;

    unordered_map<int, int> freq1;
    unordered_map<int, int> freq2;

public:
    FindSumPairs(vector<int> &nums1, vector<int> &nums2)
    {
        this->nums2 = nums2;

        for (int num : nums1)
        {
            ++freq1[num];
        }

        for (int num : nums2)
        {
            ++freq2[num];
        }

        for (const auto &pair : freq1)
        {
            keys.push_back(pair.first);
        }

        std::sort(keys.begin(), keys.end());
    }

    void add(int index, int val)
    {
        --freq2[nums2[index]];
        nums2[index] += val;
        ++freq2[nums2[index]];
    }

    int count(int tot)
    {
        int count = 0;

        for (int k : keys)
        {
            if (k >= tot)
                break;

            if (freq2.find(tot - k) != freq2.end())
            {
                count += freq1[k] * freq2[tot - k];
            }
        }

        return count;
    }
};

/**
 * Your FindSumPairs object will be instantiated and called as such:
 * FindSumPairs* obj = new FindSumPairs(nums1, nums2);
 * obj->add(index,val);
 * int param_2 = obj->count(tot);
 */

File: LeetCode/Medium/1865-Finding_Pairs_With_A_Certain_Sum/1865-FindingPairsWithACertainSum.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 */
var FindSumPairs = function (nums1, nums2) {
    this.nums2 = nums2;

    this.freq1 = new Map();
    for (const num of nums1) {
        this.freq1.set(num, (this.freq1.get(num) || 0) + 1);
    }

    this.freq2 = new Map();
    for (const num of nums2) {
        this.freq2.set(num, (this.freq2.get(num) || 0) + 1);
    }

    this.keys = Array.from(this.freq1.keys());
    this.keys.sort(function (a, b) { return a - b; });
};

/**
 * @param {number} index
 * @param {number} val
 * @return {void}
 */
FindSumPairs.prototype.add = function (index, val) {
    const oldVal = this.nums2[index];
    this.freq2.set(oldVal, this.freq2.get(oldVal) - 1);

    this.nums2[index] += val;
    const newVal = this.nums2[index];
    this.freq2.set(newVal, (this.freq2.get(newVal) || 0) + 1);
};

/**
 * @param {number} tot
 * @return {number}
 */
FindSumPairs.prototype.count = function (tot) {
    let count = 0;

    for (let k of this.keys) {
        if (k >= tot)
            break;

        if (this.freq2.has(tot - k)) {
            count += this.freq1.get(k) * this.freq2.get(tot - k);
        }
    }

    return count;
};

/**
 * Your FindSumPairs object will be instantiated and called as such:
 * var obj = new FindSumPairs(nums1, nums2)
 * obj.add(index,val)
 * var param_2 = obj.count(tot)
 */